File: ruby/Q30_subarray_sum/solution2.rb
def min_sub_array_len(target, nums)
  min_length = nums.length() + 1
  sum_list = Array.new(nums.length(), 0)

  sum_list[0] = nums[0]
  for i in 1..nums.length() - 1
    sum_list[i] = sum_list[i - 1] + nums[i]
  end

  for left in 0..nums.length() - 1
    for right in left..nums.length() - 1
      sum = sum_list[right] - sum_list[left] + nums[left]
      if sum >= target && min_length > right - left + 1
        min_length = right - left + 1
        break
      end
    end
  end

  return min_length == nums.length() + 1 ? 0 : min_length
end

File: main/Q07_two_sum/solution2.php
<?php

include 'tests.php';

function two_sum($nums, $target)
{
    $sorted = $nums;
    sort($sorted);

    $left = 0;
    $right = count($sorted) - 1;
    $answer_numbers = [];

    while ($left < $right) {
        $sum = $sorted[$left] + $sorted[$right];

        if ($sum == $target) {
            $answer_numbers = [$sorted[$left], $sorted[$right]];
            break;
        } else if ($sum < $target) {
            $left++;
        } else {
            $right--;
        }
    }

    $first_index = array_search($answer_numbers[0], $nums);
    $same = $answer_numbers[0] == $answer_numbers[1];
    $second_index = array_keys($nums, $answer_numbers[1])[$same ? 1 : 0];

    return [$first_index, $second_index];
}

// Testing
foreach ($tests as $target => $nums) {
    echo "input: [" . implode(',', $nums) . "] & " . $target . PHP_EOL
        . "answer: [" . implode(',', two_sum($nums, $target)) . "]" . PHP_EOL
        . '---------------' . PHP_EOL;
}
